Entwine Point Tile support

zarov opened the pull request:
A first view on the implementation of the [Entwine format.](entwine.io)

- feat: add Entwine Point Tile support
Based on the PointCloud basis, it is now possible to use
EntwinePointTileSource and EntwinePointTileLayer to display data from
Entwine. See https://entwine.io/ for more information on this format.
- examples: add two Entwine Point Tile examples 

I have a few things to add before review, like some documentation and the verification of the `bin` parser.

zarov committed:
feat: add Entwine Point Tile support

Based on the PointCloud basis, it is now possible to use
EntwinePointTileSource and EntwinePointTileLayer to display data from
Entwine. See https://entwine.io/ for more information on this format.

zarov committed:
examples: add two Entwine Point Tile examples

zarov commented:
@gchoqueux ready for a review !

gchoqueux commented on src/Core/EntwinePointTileNode.js:
```diff
@@ -0,0 +1,131 @@
+class EntwinePointTileNode {
+    constructor(depth, x, y, z, layer, numPoints = 0) {
+        this.layer = layer;
+        this.isEPTNode = true;
```
generally we keep the class name:         
`this.isEntwinePointTileNode = true;`

zarov replied:
Yes

gchoqueux commented on src/Core/EntwinePointTileNode.js:
```diff
@@ -0,0 +1,131 @@
+        this.layer = layer;
+        this.isEPTNode = true;
+
+        this.depth = depth;
```
could add comment to x, y, and depth

zarov replied:
Yes

gchoqueux commented on src/Core/EntwinePointTileNode.js:
```diff
@@ -0,0 +1,131 @@
+    add(node) {
+        this.children.push(node);
+        node.parent = this;
+        // TODO: read this.source.sources to get the corresponding bbox - if it
```
`this.source.sources`??

zarov replied:
I removed the part in EPTSource associated to that, so the TODO has been removed

gchoqueux commented on src/Core/EntwinePointTileNode.js:
```diff
@@ -0,0 +1,131 @@
+        node.bbox.max.copy(node.bbox.min).add(size);
+
+
+        // TODO: probleme avec le Z et les sous hierarchy
```
could add english comment and detail the issue?

zarov replied:
This is not a problem anymore, sorry about that

gchoqueux commented on src/Core/EntwinePointTileNode.js:
```diff
@@ -0,0 +1,131 @@
+        });
+    }
+
+    findCommonAncestor(node) {
```
What is difference with `PointCloudNode.findCommonAncestor()`?

zarov replied:
None anymore, thanks for seeing this

gchoqueux commented on src/Core/EntwinePointTileNode.js:
```diff
@@ -0,0 +1,131 @@
+        this.y = y;
+        this.z = z;
+
+        this.id = `${depth}-${x}-${y}-${z}`;
```
could add `buildId` methode?

zarov replied:
Yes

gchoqueux commented on src/Core/EntwinePointTileNode.js:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    findAndCreateChild(depth, x, y, z, hierarchy, stack) {
+        const id = `${depth}-${x}-${y}-${z}`;
```
could use  new `buildId` method?

zarov replied:
Yes

gchoqueux commented on src/Layer/EntwinePointTileLayer.js:
```diff
@@ -0,0 +1,68 @@
+     * layer. Default to `EPSG:4326`.
+     */
+    constructor(id, config) {
+        config.noInitialization = true;
```
what is `config.noInitialization ` ?

zarov replied:
Something I forgot to remove

gchoqueux commented on src/Provider/PointCloudProvider.js:
```diff
@@ -36,8 +36,10 @@ export default {
             points.matrixAutoUpdate = false;
-            points.position.copy(node.bbox.min);
-            points.scale.copy(layer.scale);
+            if (!layer.isEntwinePointTileLayer) {
```
why is only for `EntwinePointTileLayer`?

zarov replied:
Because EPT are always in world coordinates

gchoqueux commented on src/Source/EntwinePointTileSource.js:
```diff
@@ -0,0 +1,79 @@
+     * @constructor
+     *
+     * @param {Object} config - The configuration
+     * @param {Object} config.parserOptions - The options to pass to the parser.
```
`config.parserOptions` was removed

zarov replied:
Yes, I added the in/out system

gchoqueux replied:
nice

gchoqueux commented on src/Source/EntwinePointTileSource.js:
```diff
@@ -0,0 +1,79 @@
+            this.parse = this.dataType === 'laszip' ? data => this.parseLaz(data, config.parserOptions) : data => this.parseBin(data, config.parserOptions);
+            this.extension = this.dataType === 'laszip' ? 'laz' : 'bin';
+
+            if (this.srs && this.srs.authority && this.srs.horizontal) {
```
What is `srs`? is the projection?

zarov replied:
Yes, see here https://entwine.io/entwine-point-tile.html#srs

gchoqueux commented on src/Source/EntwinePointTileSource.js:
```diff
@@ -0,0 +1,79 @@
+            this.extension = this.dataType === 'laszip' ? 'laz' : 'bin';
+
+            if (this.srs && this.srs.authority && this.srs.horizontal) {
+                this.projection = `${this.srs.authority}:${this.srs.horizontal}`;
```
use `this.crs` instead of `this.projection`

zarov replied:
Yes

gchoqueux commented on src/Source/EntwinePointTileSource.js:
```diff
@@ -0,0 +1,79 @@
+                        this.sourcesMetadata[s] = source;
+                    });
+                }
+            });
```
Could you remove this part and add unsupported bin format comment?

zarov replied:
Yes

zarov committed:
refacto(source): use in/out system in PointCloudNode#load

Try to match the in/out system present in Source#loadData. A little bit
of work still need to be done, but this is a first step.

BREAKING CHANGE: the signature function of LASParser.parse and
PotreeBinParser.parse has changed.

zarov commented:
@gchoqueux PR updated

gchoqueux commented on src/Layer/EntwinePointTileLayer.js:
```diff
@@ -0,0 +1,69 @@
+        super(id, config);
+        this.isEntwinePointTileLayer = true;
+
+        this.source = config.source;
```
could you remove it?

zarov replied:
Yes, done
